calicoctl/calicoctl/commands/ipam/split.go

```diff
@@ -0,0 +1,208 @@
+                               YAML or JSON format.
+                               [default: ` + constants.DefaultConfigPath + `]
+     --allow-version-mismatch  Allow client and cluster versions mismatch.
+
+Description:
+  The ipam split command splits an IP pool specified by the specified CIDR into
+  the specified number of smaller IP pools. Each child IP pool will be of equal
+  size. IP pools can only be split into a number of smaller pools that is a power
+  of 2.
+
+Examples:
+  # Split the IP pool specified by 172.0.0.0/8 into 2 smaller pools
+  calicoctl ipam split 172.0.0.0/8 2
+`
+	// Replace all instances of BINARY_NAME with the name of the binary.
+	name, _ := util.NameAndDescription()
+	doc = strings.ReplaceAll(doc, "<BINARY_NAME>", name)
+
+	parsedArgs, err := docopt.ParseArgs(doc, args, "")
+	if err != nil {
+		return fmt.Errorf("Invalid option: 'calicoctl %s'. Use flag '--help' to read about a specific subcommand.", strings.Join(args, " "))
+	}
+	if len(parsedArgs) == 0 {
+		return nil
+	}
+
+	err = common.CheckVersionMismatch(parsedArgs["--config"], parsedArgs["--allow-version-mismatch"])
+	if err != nil {
+		return err
+	}
+
+	cf := parsedArgs["--config"].(string)
+	// Get the backend client.
+	client, err := clientmgr.NewClient(cf)
+	if err != nil {
+		return err
+	}
+
+	// Check that the datastore is locked
+	ctx := context.Background()
+	locked, err := common.CheckLocked(ctx, client)
+	if err != nil {
+		return fmt.Errorf("Error while checking if datastore was locked: %s", err)
+	} else if !locked {
+		return fmt.Errorf("Datastore is not locked. Run the `calicoctl datastore migrate lock` command in order split the IP pools.")
+	}
+
+	// Find the specified IP pool.
+	poolList, err := client.IPPools().List(ctx, options.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Unable to list IP pools to find the pool specified by %s", parsedArgs["<CIDR>"])
+	}
+
+	var oldPool *apiv3.IPPool
+	for _, pool := range poolList.Items {
+		if pool.Spec.CIDR == parsedArgs["<CIDR>"] {
+			oldPool = &pool
+		}
+	}
+
+	if oldPool == nil {
+		return fmt.Errorf("Unable to find IP pool covering the specified CIDR %s", parsedArgs["<CIDR>"])
+	}
+
+	// Disable the specified IP pool.
+	oldPool.Spec.Disabled = true
+	oldPool, err = client.IPPools().Update(ctx, oldPool, options.SetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error disabling IP pool %s", parsedArgs["<CIDR>"])
+	}
+
+	// Calculate the split pool CIDRs.
+	numString := parsedArgs["<NUMBER>"].(string)
+	splitNum, err := strconv.Atoi(numString)
+	if err != nil {
+		return fmt.Errorf("Error reading number to split IP pools into. %s is not a valid number: %v", numString, err)
+	}
+
+	splitCIDRs, err := splitCIDR(oldPool.Spec.CIDR, splitNum)
+	if err != nil {
+		return fmt.Errorf("Error splitting the CIDR %s into %d CIDRs: %v", oldPool.Spec.CIDR, splitNum, err)
+	}
+
+	// Create the new split pools using UnsafeCreate.
+	for i, cidr := range splitCIDRs {
+		splitPool := &apiv3.IPPool{
```

Comment:
I think we need to use a DeepCopy to make a copy here, rather than manually copying fields across. This is fragile when new fields are added, and already doesn't include everything we might want (e.g., labels and annotations from the old pool).

So we probably need:

```
newP := oldPool.DeepCopyObject()
newP.Name = ...
newP.Spec.CIDR = ...
```